File: torrentplugin/app/src/main/java/com/torrentplugin/scrapers/TorrentSearch.kt
package com.torrentplugin.scrapers

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Runs every active scraper in parallel, flattens and dedupes by infohash, and sorts by
 * seeders. Each scraper gets its own timeout so one slow site can't hold up the rest - a
 * partial result set is still returned.
 */
class ScraperAggregator {

    // UIndex reliably returns HTTP 403 (bot-blocked), so it's left out of the active set.
    private val scrapers = listOf(KnabenScraper(), ThePirateBayScraper())

    suspend fun searchAll(query: String): List<ScrapedTorrent> = coroutineScope {
        val perScraper = scrapers.map { scraper ->
            async(Dispatchers.IO) {
                withTimeoutOrNull(15_000) {
                    try { scraper.search(query) }
                    catch (e: Exception) { Log.w(TAG, "${scraper.source} threw: ${e.message}"); emptyList() }
                } ?: run { Log.w(TAG, "${scraper.source} timed out"); emptyList() }
            }
        }.awaitAll().flatten()
        Log.i(TAG, "Aggregator: '$query' -> ${perScraper.size} raw results before dedupe")

        val byHash = LinkedHashMap<String, ScrapedTorrent>()
        for (t in perScraper) {
            val key = infohashOf(t.magnet) ?: t.magnet
            // Keep whichever copy reports more seeders.
            val existing = byHash[key]
            if (existing == null || seedersInt(t) > seedersInt(existing)) byHash[key] = t
        }
        byHash.values.sortedByDescending { seedersInt(it) }
    }

    companion object {
        private const val TAG = "TorrentScraper"
        private val BTIH = Regex("btih:([A-Fa-f0-9]{40}|[A-Za-z2-7]{32})")
        fun infohashOf(magnet: String): String? =
            BTIH.find(magnet)?.groupValues?.get(1)?.lowercase()
        fun seedersInt(t: ScrapedTorrent): Int = t.seeders.replace(",", "").trim().toIntOrNull() ?: 0
    }
}

/**
 * The title/quality helpers PlayTorrioV2 keeps in its TorrentFilter, reduced to what the
 * movie path needs: a loose title match (so unrelated results are dropped) and a quality tag
 * pulled out of the release name for display/sorting.
 */
object TorrentFilter {

    private val QUALITY_TAGS = listOf(
        "2160p" to "4K", "4k" to "4K", "uhd" to "4K",
        "1080p" to "1080p", "720p" to "720p", "480p" to "480p"
    )

    fun qualityOf(name: String): String? {
        val lower = name.lowercase()
        return QUALITY_TAGS.firstOrNull { lower.contains(it.first) }?.second
    }

    fun qualityRank(quality: String?): Int = when (quality) {
        "4K" -> 400; "1080p" -> 300; "720p" -> 200; "480p" -> 100; else -> 0
    }

    /** Parses a size string like "1.85 GiB" / "700 MB" / "5.2 GB" to bytes, or null if unknown. */
    fun sizeBytes(size: String): Long? {
        val m = Regex("([\\d.,]+)\\s*(TB|TiB|GB|GiB|MB|MiB|KB|KiB|B)", RegexOption.IGNORE_CASE)
            .find(size) ?: return null
        val value = m.groupValues[1].replace(",", "").toDoubleOrNull() ?: return null
        val mult = when (m.groupValues[2].uppercase()) {
            "TB", "TIB" -> 1024.0 * 1024 * 1024 * 1024
            "GB", "GIB" -> 1024.0 * 1024 * 1024
            "MB", "MIB" -> 1024.0 * 1024
            "KB", "KIB" -> 1024.0
            else -> 1.0
        }
        return (value * mult).toLong()
    }

    /** True if the torrent is small enough to stream on a low-storage device (or size unknown). */
    fun withinSizeLimit(size: String): Boolean {
        val bytes = sizeBytes(size) ?: return true // keep unparseable sizes rather than hide them
        return bytes <= MAX_SIZE_BYTES
    }

    /** 8 GB ceiling - larger rips are impractical to stream/cache on the target boxes. */
    private const val MAX_SIZE_BYTES = 8L * 1024 * 1024 * 1024

    /** Normalizes a title for comparison: lowercase, punctuation to spaces, collapse spaces. */
    fun normalize(title: String): String =
        title.lowercase().replace(Regex("[^a-z0-9]+"), " ").trim()

    /** True if the release name looks like the given season/episode (SxxEyy or NxMM forms). */
    fun matchesEpisode(resultName: String, season: Int, episode: Int): Boolean {
        val n = resultName.lowercase()
        val forms = listOf(
            String.format(java.util.Locale.ROOT, "s%02de%02d", season, episode),
            String.format(java.util.Locale.ROOT, "%dx%02d", season, episode),
            "s${season}e${episode}"
        )
        return forms.any { n.contains(it) }
    }

    /** A movie result is kept if its normalized name contains every word of the wanted title. */
    fun matchesMovie(resultName: String, wantedTitle: String): Boolean {
        val hay = normalize(resultName)
        val words = normalize(wantedTitle).split(' ').filter { it.length > 1 }
        if (words.isEmpty()) return true
        return words.all { hay.contains(it) }
    }
}

File: scripts/torrent-search.js
// stream_search plugin: searches public torrent indexers (ThePirateBay, Knaben) for a title.
// Ported from the old torrentplugin APK's Scrapers.kt + TorrentSearch.kt (search half only -
// resolveNatively means Lumora's own com.lumora.torrent.TorrentEngine handles resolve(), never
// calling back into this script; JS can't run a persistent local HTTP server or hold a libtorrent
// session open, only the host app can) onto Lumora's in-process JS plugin host API - see
// JsPluginContract.kt.

PLUGIN = {
    id: "torrent.search",
    label: "Torrent Search",
    description: "Searches public torrent indexers for a title and streams the result via Lumora's built-in torrent engine.",
    capabilities: ["stream_search"],
    resolvesNatively: true,
};

var MAX_RESULTS = 200;
var MAX_SIZE_BYTES = 8 * 1024 * 1024 * 1024; // 8 GB ceiling - impractical to stream/cache above this
var UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";
var SCRAPE_HEADERS = {
    "User-Agent": UA,
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
};

// ── Scrapers ──

/** ThePirateBay mirror. Paginated table; stops on the first empty page. */
function tpbSearch(query) {
    var base = "https://1.piratebays.to";
    var maxPages = 10;
    var out = [];
    var encoded = encodeURIComponent(query);
    for (var page = 1; page <= maxPages; page++) {
        var url = page === 1
            ? base + "/s/?q=" + encoded + "&video=on&category=0"
            : base + "/s/page/" + page + "/?q=" + encoded + "&video=on&category=0";
        var html;
        try {
            var resp = host.httpGet(url, SCRAPE_HEADERS);
            if (resp.status < 200 || resp.status >= 300) break;
            html = resp.body;
        } catch (e) {
            break;
        }

        var rows = host.selectAll(html, "table tr");
        var added = 0;
        for (var i = 0; i < rows.length; i++) {
            var row = rows[i];
            if (host.selectFirst(row, "th")) continue; // header row
            var name = host.selectText(row, "a.detLink");
            if (!name) continue;
            var magnet = host.selectAttr(row, "a[href^=magnet:]", "href");
            if (!magnet) continue;
            out.push({
                name: name,
                magnet: magnet,
                size: host.selectTextAt(row, "td", 4) || "Unknown",
                seeders: host.selectTextAt(row, "td", 5) || "Unknown",
                source: "ThePirateBay",
            });
            added++;
        }
        if (added === 0) break;
    }
    return out;
}

/** Knaben. Single path-style page, pre-sorted by seeders. */
function knabenSearch(query) {
    var base = "https://knaben.org";
    var url = base + "/search/" + encodeURIComponent(query) + "/0/1/seeders";
    var html;
    try {
        var resp = host.httpGet(url, SCRAPE_HEADERS);
        if (resp.status < 200 || resp.status >= 300) return [];
        html = resp.body;
    } catch (e) {
        return [];
    }

    var out = [];
    var rows = host.selectAll(html, "tbody tr");
    for (var i = 0; i < rows.length; i++) {
        var row = rows[i];
        var titleCellHtml = host.selectFirst(row, "td.text-wrap");
        if (!titleCellHtml) continue;
        var magnet = host.selectAttr(titleCellHtml, "a[href^=magnet:]", "href");
        if (!magnet) continue;
        var name = host.selectAttr(titleCellHtml, "a[href^=magnet:]", "title")
            || host.selectText(titleCellHtml, "a[href^=magnet:]")
            || "Unknown";

        var allCells = host.selectAll(row, "td");
        var titleIdx = -1;
        for (var c = 0; c < allCells.length; c++) {
            if (allCells[c] === titleCellHtml) {
                titleIdx = c;
                break;
            }
        }
        var size = titleIdx >= 0 && titleIdx + 1 < allCells.length ? host.textOf(allCells[titleIdx + 1]) : "";
        var seeders = allCells.length >= 3 ? host.textOf(allCells[allCells.length - 3]) : "";

        out.push({
            name: name,
            magnet: magnet,
            size: size || "Unknown",
            seeders: seeders || "Unknown",
            source: "Knaben",
        });
    }
    return out;
}

// ── Aggregation (dedupe by infohash, keep the higher-seeders copy, sort by seeders) ──

var BTIH_REGEX = /btih:([A-Fa-f0-9]{40}|[A-Za-z2-7]{32})/;

function infohashOf(magnet) {
    var m = BTIH_REGEX.exec(magnet);
    return m ? m[1].toLowerCase() : null;
}

function seedersInt(t) {
    var n = parseInt(String(t.seeders).replace(/,/g, "").trim(), 10);
    return isNaN(n) ? 0 : n;
}

function searchAll(query) {
    var all = [].concat(knabenSearch(query), tpbSearch(query));
    var byHash = {};
    var order = [];
    for (var i = 0; i < all.length; i++) {
        var t = all[i];
        var key = infohashOf(t.magnet) || t.magnet;
        var existing = byHash[key];
        if (!existing || seedersInt(t) > seedersInt(existing)) {
            if (!existing) order.push(key);
            byHash[key] = t;
        }
    }
    var results = order.map(function (key) { return byHash[key]; });
    results.sort(function (a, b) { return seedersInt(b) - seedersInt(a); });
    return results;
}

// ── Filtering (mirrors TorrentFilter) ──

var QUALITY_TAGS = [["2160p", "4K"], ["4k", "4K"], ["uhd", "4K"], ["1080p", "1080p"], ["720p", "720p"], ["480p", "480p"]];

function qualityOf(name) {
    var lower = name.toLowerCase();
    for (var i = 0; i < QUALITY_TAGS.length; i++) {
        if (lower.indexOf(QUALITY_TAGS[i][0]) !== -1) return QUALITY_TAGS[i][1];
    }
    return null;
}

function sizeBytes(size) {
    var m = /([\d.,]+)\s*(TB|TiB|GB|GiB|MB|MiB|KB|KiB|B)/i.exec(size);
    if (!m) return null;
    var value = parseFloat(m[1].replace(/,/g, ""));
    if (isNaN(value)) return null;
    var unit = m[2].toUpperCase();
    var mult = 1;
    if (unit === "TB" || unit === "TIB") mult = 1024 * 1024 * 1024 * 1024;
    else if (unit === "GB" || unit === "GIB") mult = 1024 * 1024 * 1024;
    else if (unit === "MB" || unit === "MIB") mult = 1024 * 1024;
    else if (unit === "KB" || unit === "KIB") mult = 1024;
    return Math.floor(value * mult);
}

function withinSizeLimit(size) {
    var bytes = sizeBytes(size);
    return bytes === null || bytes <= MAX_SIZE_BYTES; // keep unparseable sizes rather than hide them
}

function normalize(title) {
    return title.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
}

function matchesEpisode(resultName, season, episode) {
    var n = resultName.toLowerCase();
    var pad2 = function (v) { return (v < 10 ? "0" : "") + v; };
    var forms = ["s" + pad2(season) + "e" + pad2(episode), season + "x" + pad2(episode), "s" + season + "e" + episode];
    for (var i = 0; i < forms.length; i++) {
        if (n.indexOf(forms[i]) !== -1) return true;
    }
    return false;
}

function matchesMovie(resultName, wantedTitle) {
    var hay = normalize(resultName);
    var words = normalize(wantedTitle).split(" ").filter(function (w) { return w.length > 1; });
    if (words.length === 0) return true;
    return words.every(function (w) { return hay.indexOf(w) !== -1; });
}

/**
 * Strips catalog cruft that would never appear in a release name: a short provider/category tag
 * prefix like "4K-D+ - " or "D+ - ", and a bracketed year (the year is sent separately). Kept
 * conservative so real dashed titles ("Mission: Impossible - ...") survive.
 */
function cleanTitle(raw) {
    return raw
        .replace(/^[\w+-]{1,6}\s-\s/, "")
        .replace(/\(\d{4}\)/, " ")
        .replace(/\s+/g, " ")
        .trim();
}

function pad2(n) {
    return (n < 10 ? "0" : "") + n;
}

// ── stream_search entry point ──

function search(host, query, year, season, episode) {
    var title = cleanTitle((query || "").trim());
    if (!title) throw new Error("Empty search query");

    var epTag = season && episode ? "S" + pad2(season) + "E" + pad2(episode) : null;
    // For an episode search append SxxEyy (and drop the year, which rarely appears in episode
    // release names); otherwise a movie search keeps the year to disambiguate remakes.
    var searchQuery = epTag ? title + " " + epTag : year ? title + " " + year : title;

    host.reportProgress("Searching torrent sites");
    var raw = searchAll(searchQuery);
    var filtered = raw.filter(function (t) {
        return matchesMovie(t.name, title)
            && withinSizeLimit(t.size)
            && (!epTag || matchesEpisode(t.name, season, episode));
    });

    var sent = 0;
    for (var i = 0; i < filtered.length && sent < MAX_RESULTS; i++) {
        var t = filtered[i];
        host.reportResult({
            title: t.name,
            token: t.magnet, // resolvesNatively - the host resolves a magnet token itself
            seeders: seedersInt(t),
            size: t.size,
            quality: qualityOf(t.name),
            source: t.source,
        });
        sent++;
    }

    return sent === 0 ? "No torrents found" : sent + " result" + (sent === 1 ? "" : "s");
}
